File: Tools/ParseBlueprintLog.py
#!/usr/bin/env python3
"""Parse EMOJ simple.log batch-test output into JSON index."""
import json
import re
import sys
from pathlib import Path

LOG_PATH = Path.home() / "Documents/My Games/Terraria/tModLoader/Logs/EMOJ/2026-05-27_12-28-08/simple.log"
OUT_DIR = Path(__file__).resolve().parent.parent / "Issues" / "audit_data"

SLOTS = [
    "Block", "Wall", "Bathtub", "Bed", "Bookcase", "Candelabra", "Candle", "Chandelier",
    "Chair", "Chest", "Clock", "Door", "Dresser", "Lamp", "Lantern", "Piano", "Platform",
    "Sink", "Sofa", "Table", "Toilet", "Workbench",
]

SEED_RE = re.compile(
    r"batch-test seed=(\d+)\s+name=(.+?)\s+material=(\d+)\s+wiki=(\d+)/22\s+"
    r"accuracy=(\d+)/(\d+)\s+mismatch=(\d+)\s+lineage_miss=(\d+)\s+vanilla_leak=(\d+)"
)
SLOT_RE = re.compile(
    r"^\s+(Block|Wall|Bathtub|Bed|Bookcase|Candelabra|Candle|Chandelier|"
    r"Chair|Chest|Clock|Door|Dresser|Lamp|Lantern|Piano|Platform|"
    r"Sink|Sofa|Table|Toilet|Workbench)=(\d+)\|([^|]*)\|(.*)$"
)
BLUEPRINT_RE = re.compile(r"\[Blueprint\]\s*(.+)$")


def parse_log(path: Path) -> dict:
    entries = {}
    current = None
    current_seed = None
    in_slots = False

    def flush():
        nonlocal current, current_seed
        if current and len(current["slots"]) >= 20:
            entries[str(current_seed)] = current

    text = path.read_text(encoding="utf-8", errors="replace")
    for raw in text.splitlines():
        m = BLUEPRINT_RE.search(raw)
        line = m.group(1) if m else raw

        sm = SEED_RE.search(line)
        if sm:
            flush()
            current_seed = int(sm.group(1))
            name = sm.group(2)
            name = re.sub(r"\s+(incomplete|candidates=\d+).*$", "", name).strip()
            current = {
                "seed": current_seed,
                "name": name,
                "material": int(sm.group(3)),
                "wiki_filled": int(sm.group(4)),
                "accuracy": int(sm.group(5)),
                "accuracy_denom": int(sm.group(6)),
                "slot_mismatch": int(sm.group(7)),
                "lineage_miss": int(sm.group(8)),
                "vanilla_leak": int(sm.group(9)),
                "incomplete": "incomplete" in raw,
                "slots": {},
            }
            in_slots = False
            continue

        if line.startswith("batch-test slots seed="):
            sid = int(line.split("seed=")[1].split()[0])
            in_slots = sid == current_seed
            continue

        if in_slots and current:
            tm = SLOT_RE.match(line)
            if tm:
                slot = tm.group(1)
                current["slots"][slot] = {
                    "type": int(tm.group(2)),
                    "internal": tm.group(3),
                    "display": tm.group(4),
                }
            elif not line.strip().startswith("Slot="):
                flush()
                in_slots = False

    flush()
    return entries


def main():
    log_path = Path(sys.argv[1]) if len(sys.argv) > 1 else LOG_PATH
    out_dir = Path(sys.argv[2]) if len(sys.argv) > 2 else OUT_DIR
    out_dir.mkdir(parents=True, exist_ok=True)

    entries = parse_log(log_path)
    index_path = out_dir / "batch_slots_index.json"
    index_path.write_text(json.dumps(entries, ensure_ascii=False, indent=2), encoding="utf-8")

    all_entries = list(entries.values())
    scored = [e for e in all_entries if e["accuracy_denom"] > 0]
    high_lineage = sorted(
        [e for e in scored if e["lineage_miss"] >= 5],
        key=lambda e: (-e["lineage_miss"], e["seed"]),
    )[:30]

    summary = {
        "total_seeds": len(all_entries),
        "scored": len(scored),
        "avg_wiki": round(sum(e["wiki_filled"] for e in all_entries) / max(len(all_entries), 1), 2),
        "total_lineage_miss": sum(e["lineage_miss"] for e in scored),
        "high_lineage": [
            {k: e[k] for k in ("seed", "name", "material", "wiki_filled", "lineage_miss", "accuracy", "accuracy_denom")}
            for e in high_lineage
        ],
    }
    (out_dir / "batch_summary.json").write_text(
        json.dumps(summary, ensure_ascii=False, indent=2), encoding="utf-8"
    )
    print(f"Parsed {len(all_entries)} seeds -> {index_path}")


if __name__ == "__main__":
    main()

File: Tools/lib/parseBlueprintLog.js
/** Parse EMOJ simple.log batch-test output into JSON index. */
const BLUEPRINT_RE = /\[Blueprint\]\s*(.+)$/;

const SEED_RE =
  /batch-test seed=(\d+)\s+name=(.+?)\s+material=(\d+)\s+wiki=(\d+)\/22\s+accuracy=(\d+)\/(\d+)\s+mismatch=(\d+)\s+lineage_miss=(\d+)\s+vanilla_leak=(\d+)(?:\s+material_miss=(\d+))?(?:\s+style_slot_match=(\d+)\/(\d+))?(?:\s+set_conf=(\w+)\s+style_align=(\d+)%)?(?:\s+candidates=(\d+))?/;

const SLOT_RE =
  /^\s*(Block|Wall|Bathtub|Bed|Bookcase|Candelabra|Candle|Chandelier|Chair|Chest|Clock|Door|Dresser|Lamp|Lantern|Piano|Platform|Sink|Sofa|Table|Toilet|Workbench)=(\d+)\|([^|]*)\|(.*)$/;

function parseBlueprintLog(filePath) {
  const fs = require('fs');
  const text = fs.readFileSync(filePath, 'utf8');
  const entries = {};
  let current = null;
  let currentSeed = null;
  let inSlots = false;

  const flush = () => {
    if (current && Object.keys(current.slots).length >= 20) {
      entries[String(currentSeed)] = current;
    }
  };

  for (const raw of text.split(/\r?\n/)) {
    const bm = raw.match(BLUEPRINT_RE);
    const line = bm ? bm[1] : raw;

    const sm = line.match(SEED_RE);
    if (sm) {
      flush();
      currentSeed = parseInt(sm[1], 10);
      let name = sm[2].replace(/\s+(incomplete|candidates=\d+).*$/, '').trim();
      current = {
        seed: currentSeed,
        name,
        material: parseInt(sm[3], 10),
        wiki_filled: parseInt(sm[4], 10),
        accuracy: parseInt(sm[5], 10),
        accuracy_denom: parseInt(sm[6], 10),
        slot_mismatch: parseInt(sm[7], 10),
        lineage_miss: parseInt(sm[8], 10),
        vanilla_leak: parseInt(sm[9], 10),
        material_miss: sm[10] ? parseInt(sm[10], 10) : 0,
        style_slot_match: sm[11] ? parseInt(sm[11], 10) : 0,
        style_slot_filled: sm[12] ? parseInt(sm[12], 10) : 0,
        set_conf: sm[13] || null,
        style_align: sm[14] ? parseInt(sm[14], 10) : null,
        candidates: sm[15] ? parseInt(sm[15], 10) : null,
        incomplete: raw.includes('incomplete'),
        slots: {},
      };
      inSlots = false;
      continue;
    }

    if (line.startsWith('batch-test slots seed=')) {
      const sid = parseInt(line.split('seed=')[1].split(/\s/)[0], 10);
      inSlots = sid === currentSeed;
      continue;
    }

    if (inSlots && current) {
      const tm = line.match(SLOT_RE);
      if (tm) {
        current.slots[tm[1]] = {
          type: parseInt(tm[2], 10),
          internal: tm[3],
          display: tm[4],
        };
      } else if (
        !line.trim().startsWith('Slot=') &&
        !line.startsWith('batch-test seed=') &&
        !line.startsWith('batch-test slots seed=')
      ) {
        if (Object.keys(current.slots).length >= 20) {
          flush();
        }
        inSlots = false;
      }
    }
  }
  flush();
  return entries;
}

module.exports = { parseBlueprintLog, SEED_RE, SLOT_RE };
